### PR TITLE
Templating

### PR DESCRIPTION
**Pull request by dlip**
I decided to revert to standard bash logic of subsequent variables override previous ones. I have added templating so you can use `{{CI_VERSION}}` etc to set the version from CI/CD.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -118,13 +145,30 @@ VERSION=v1.0.0
...
+
+To write a literal `{{`, escape it by putting it between curly braces `{{{{}}`
+
 ## Variable Precedence
 
-- Existing environment variables will not be overridden
-- Includes will override variables which are declared afterwards, if you have a shared common.env include add it to the bottom of your file so preceding variables can override its contents
+- Variables are processed top to bottom and lower varibles will overrite previous ones
+- Includes will be evaluated in the line they are written. If you have a shared common.env you can include it at the top of your file so following variables can override its values
+- Output will be sorted alphabetically at the end after the above evaluations for consistency
```
Comment (leontee) [inline README.md]: Had to read that line point about sorting a couple of times, maybe reword it so it's clear that all lines generated by envy will be sorted alphabetically before the output is written. I think the 'after the above evaluations' throws it off a bit.

This and a nitpick at 'overrite' to 'overwrite'

File: envy.sh [hunk 2]
```diff
@@ -20,37 +79,45 @@ envy() {
         CONTENTS=$(cat "${FILENAME}" | grep "^[^#]")
     fi
     while read -r PAIR; do
-        K=$(echo ${PAIR} | sed 's/\([^=]*\)=\(.*\)/\1/')
-        V=$(echo ${PAIR} | sed 's/\([^=]*\)=\(.*\)/\2/')
+        K=$(sed 's/\([^=]*\)=\(.*\)/\1/' <<< "${PAIR}")
+        V=$(sed 's/\([^=]*\)=\(.*\)/\2/' <<< "${PAIR}")
+        # Check if templating
+        if grep -q "{{.*}}" <<< "${V}"; then
```
Comment (leontee) [inline envy.sh:85]: Seeing quite a few references to .* representing the variable token between the {{ and }} delimiters. Does this mean that someone could potentially have {{}} in the template?
Maybe references could be more specific to be something like `[a-zA-Z0-9-_]\+` or whatever the valid definition of a variable is :)

File: envy.sh [hunk 3]
```diff
@@ -20,37 +79,45 @@ envy() {
...
-                    echo "${BASH_FORMAT}"
-                elif [ "${OUTPUT}" == "env-file" ]; then
-                    echo "${PAIR}"
-                elif [ "${OUTPUT}" == "make" ]; then
-                    MAKE_ESCAPED_VALUE=$(sed 's/\([$]\)/$\1/g' <<< "${V}")
-                    MAKE_ESCAPED_VALUE=$(sed 's/\([#\\]\)/\\\1/g' <<< "${MAKE_ESCAPED_VALUE}")
-                    echo "export ${K}=${MAKE_ESCAPED_VALUE}"
-                fi
-            fi
+            export ${K}="$(bash_escape "${V}")"
```
Comment (leontee) [inline envy.sh:91]: Thought this should be failing given the 2 sets of double quotes, are we not exercising this in tests?

File: envy.sh [hunk 4]
```diff
@@ -20,37 +79,45 @@ envy() {
...
+        V=$(printenv "${ENVY_NAMESPACE}${K}")
+        if [ "${OUTPUT}" == "bash" ]; then
+            echo "export ${K}=$(bash_escape "${V}")"
+        elif [ "${OUTPUT}" == "env-file" ]; then
+            echo "${K}=${V}"
+        elif [ "${OUTPUT}" == "make" ]; then
+            MAKE_ESCAPED_VALUE=$(sed 's/\([$]\)/$\1/g' <<< "${V}")
+            MAKE_ESCAPED_VALUE=$(sed 's/\([#\\]\)/\\\1/g' <<< "${MAKE_ESCAPED_VALUE}")
+            echo "export ${K}=${MAKE_ESCAPED_VALUE}"
+        fi
```
Comment (leontee) [inline envy.sh:115]: Would be good to have an 'else' here to handle a user mistake in inputs

**Review by leontee — CHANGES_REQUESTED** [review]

File: envy.sh (hunk 2 above)
Comment (dlip) [reply envy.sh:85]: This is just a basic check to see if there is any templating logic to know if it should even bother to run the templating

File: envy.sh (hunk 3 above)
Comment (dlip) [reply envy.sh:91]: It's definitely exercised and seems to be parsed correctly by bash, i guess the $() is evaluated as a separate expression since its in the subshell

**Comment by dlip** [comment]
@leontee Updated

File: envy.sh (hunk 2 above)
Comment (leontee) [reply envy.sh:85]: Understood, although I meant for my comment to be more generic referring to the '.*' across the board for things like pattern matching or substitution e.g. `sed 's/\([^=]*\)=\(.*\)/\1/')`. Is there any reason why we shouldn't consider replacing these with something more concrete that enforces 1..* characters (e.g. `.\+`) rather than 0..* (`.*`)?

Comment (dlip) [reply envy.sh:85]: Using Bash parsing like in mush
```
K=${PAIR%%"="*}
V=${PAIR#*"="*}
```

Comment (leontee) [reply envy.sh:85]: 👍

**Review by leontee — APPROVED** [review]
Nice pr!
